FactoryTest/FactoryTest.ino

```diff
@@ -0,0 +1,342 @@
+
+/********************************************************************
+ * Fan
+ ********************************************************************/
+
+// How many milliseconds should there be between readings of the fan speed. A smaller value will update
+// more often, while a higher value will give more accurate and smooth readings.
+const int FAN_SPEED_READING_INTERVAL = 1000;
+
+FanController fanController(FAN_RPM_PIN, FAN_SPEED_READING_INTERVAL, FAN_PWM_PIN);
+
+const int FAN_DUTYCYCLE_MINIMUM = 0;
+const int FAN_DUTYCYCLE_MEDIUM = 50;
+const int FAN_DUTYCYCLE_MAXIMUM = 100;
+
+void exerciseFan(int dutyCycle) {
+    fanController.setDutyCycle(dutyCycle);
+}
+
+void exerciseFanSpeed() {
+    fanController.setDutyCycle(FAN_DUTYCYCLE_MEDIUM);
+
+    // wait for the fan to get up to speed
+    delay(DELAY_1Sec);
+
+    // read the speed
+    unsigned int speed = fanController.getSpeed();
+
+    if (speed > 65535) {
+        // speed is too big for writeNumberToLights
+        exerciseAllLEDs();
+    } else {
+        writeNumberToLights(speed);
+    }
+
+    fanController.setDutyCycle(FAN_DUTYCYCLE_MINIMUM);
+}
+
+/********************************************************************
+ * Other Devices
+ ********************************************************************/
+
+void exerciseBuzzer() {
+    for (int i = 0; i < 3; i += 1) {
+        analogWrite(BUZZER_PIN, 255);
+        delay(DELAY_500ms);
+        analogWrite(BUZZER_PIN, 0);
+        delay(DELAY_500ms);
+    }
+}
+
+void exerciseVibrator() {
+    for (int i = 0; i < 3; i += 1) {
+        digitalWrite(VIBRATOR_PIN, HIGH);
+        delay(DELAY_500ms);
+        digitalWrite(VIBRATOR_PIN, LOW);
+        delay(DELAY_500ms);
+    }
+}
+
+void exerciseBatteryVoltage() {
+    // The battery readings we expect for minimum and maximum battery voltages.
+    // TODO: we probably need to get rid of these hard-coded constants, and use a better way to derive these numbers.
+    const int readingAt12Volts = 386;
+    const int readingAt24Volts = 784;
+
+    // For the next 10 seconds we will display the battery voltage on the LEDs.
+    // Empty battery = 1 LEDs. Full battery = 7 LEDs.
+    // As you change the input voltage, the LEDs will update accordingly.
+    unsigned long startTime = millis();
+    while (millis() - startTime < 10000) {
+        // Read the current battery voltage and limit the value to the expected range.
+        uint16_t reading = analogRead(BATTERY_VOLTAGE_PIN);
+        if (reading < readingAt12Volts) reading = readingAt12Volts;
+        if (reading > readingAt24Volts) reading = readingAt24Volts;
+
+        // Calculate how full the battery is. This will be a number between 0 and 1.
+        // (Note: we use floating point because it's easier than trying to do this in fixed point. The program memory impact appears to be negligible.)
+        float fullness = float(reading - readingAt12Volts) / float(readingAt24Volts - readingAt12Volts);
```

Comment:
we will of course want this in a library function at some point. May make sense to start derriving some of these thigns.

Reply:
I created a library function in PAPRHwDefs.h. I found some code at https://github.com/airtoall/papr-firmware/blob/752f2db06e0bd756dd24af65175921b3bed734dd/ControllerFirmware/ControllerFirmware.ino#L196 that maybe is a step in the right direction, but that code doesn't eliminate the need for hard-coded magic numbers. Let's chat about this some time.